ordinary_data/valve/tr_valve_pipe.sql

```diff
@@ -67,23 +119,49 @@ COMMENT ON TRIGGER tr_valve_pipe_delete ON qwat_od.pipe IS 'Trigger: when deleti
 
 
 /* AFTER UPDATE OF A VALVE: REEVALUATE OLD AND NEW PIPES */
-CREATE OR REPLACE FUNCTION qwat_od.ft_valve_update() RETURNS TRIGGER AS
+CREATE FUNCTION qwat_od.ft_valve() RETURNS TRIGGER AS
 $BODY$
 	BEGIN
-		PERFORM qwat_od.fn_pipe_update_valve(OLD.id);
 		PERFORM qwat_od.fn_pipe_update_valve(NEW.id);
 		RETURN NEW;
 	END;
 $BODY$
 LANGUAGE plpgsql;
-COMMENT ON FUNCTION qwat_od.ft_valve_update() IS 'Trigger: when updating a valve, reevaluate old and new pipes.';
 
-CREATE TRIGGER tr_valve_update_trigger
-	AFTER UPDATE ON qwat_od.valve
+COMMENT ON FUNCTION qwat_od.ft_valve() IS 'Trigger: when updating a valve, reevaluate old and new pipes for number of valves on them.';
+
+CREATE TRIGGER tr_valve_trigger
+	AFTER INSERT
+	ON qwat_od.valve
 	FOR EACH ROW
-	EXECUTE PROCEDURE qwat_od.ft_valve_update();
-COMMENT ON TRIGGER tr_valve_update_trigger ON qwat_od.valve IS 'Trigger: when updating a valve, reevaluate old and new pipes.';
+		EXECUTE PROCEDURE qwat_od.ft_valve();
+
+COMMENT ON TRIGGER tr_valve_trigger ON qwat_od.valve 
+IS 'Trigger: when inserting or updating a valve, reevaluate new pipes for number of valves on them.';
+
+CREATE OR REPLACE FUNCTION qwat_od.ft_before_valve()
+  RETURNS trigger AS
+$BODY$
+	BEGIN
+		IF TG_OP = 'DELETE' THEN
+			PERFORM qwat_od.fn_pipe_delete_valve(OLD.id);
+		END IF;
+		
+		RETURN OLD;
+	END;
+$BODY$
+LANGUAGE plpgsql;
+
+COMMENT ON FUNCTION qwat_od.ft_before_valve() IS 'Trigger: when deleting a valve, reevaluate pipes for number of valves on them.';
+
+CREATE TRIGGER tr_valve_delete_trigger
+  BEFORE DELETE
```

Comment:
so, my question was mainly why don't you do an AFTER DELETE here?

Reply:
I don't understand why, but if it's not done BEFORE, the update is not done.

Reply:
OK Seems related to other trigger.